modules/python/src2/cv2.cpp

```diff
@@ -159,6 +165,155 @@ catch (const cv::Exception &e) \
+template<>
+NPY_TYPES asNumpyType<float>()
+{
+    return NPY_FLOAT;
+}
+
+template<>
+NPY_TYPES asNumpyType<double>()
+{
+    return NPY_DOUBLE;
+}
+
+template <class T>
+PyArray_Descr* getNumpyTypeDescriptor()
+{
+    return PyArray_DescrFromType(asNumpyType<T>());
+}
+
+template <>
+PyArray_Descr* getNumpyTypeDescriptor<size_t>()
+{
+#if SIZE_MAX == ULONG_MAX
+    return PyArray_DescrFromType(NPY_ULONG);
+#elif SIZE_MAX == ULLONG_MAX
+    return PyArray_DescrFromType(NPY_ULONGLONG);
+#else
+    return PyArray_DescrFromType(NPY_UINT);
+#endif
+}
+
+template <class T, class U>
+bool isRepresentable(U value) {
+    return (std::numeric_limits<T>::min() <= value) && (value <= std::numeric_limits<T>::max());
+}
+
+template<class T>
+typename stdx::enable_if<stdx::is_floating_point<T>::value, bool>::type
+canBeSafelyCasted(PyObject* obj, PyArray_Descr* to)
+{
+    return PyArray_CanCastTo(PyArray_DescrFromScalar(obj), to) != 0;
+}
+
+
+template<class T>
+typename stdx::enable_if<stdx::is_integral<T>::value && stdx::is_signed<T>::value, bool>::type
+canBeSafelyCasted(PyObject* obj, PyArray_Descr* to)
+{
+    PyArray_Descr* from = PyArray_DescrFromScalar(obj);
+    if (PyArray_CanCastTo(from, to))
+    {
+        return true;
+    }
+    else
+    {
+        // False negative scenarios:
+        // - Input has wider limits but value is representable within output limits
+        int64_t input = 0;
+        PyArray_CastScalarToCtype(obj, &input, getNumpyTypeDescriptor<int64_t>());
+        return isRepresentable<T>(static_cast<int64_t>(input));
+    }
+}
+
+template<class T>
+typename stdx::enable_if<stdx::is_integral<T>::value && stdx::is_unsigned<T>::value, bool>::type
+canBeSafelyCasted(PyObject* obj, PyArray_Descr* to)
+{
+    PyArray_Descr* from = PyArray_DescrFromScalar(obj);
+    if (PyArray_CanCastTo(from, to))
+    {
+        return true;
+    }
+    else
+    {
+        // False negative scenarios:
+        // - Signed input is positive so it can be safely cast to unsigned output
+        // - Input has wider limits but value is representable within output limits
+        // - All the above
+        if (PyDataType_ISSIGNED(from))
+        {
+            int64_t input = 0;
+            PyArray_CastScalarToCtype(obj, &input, getNumpyTypeDescriptor<int64_t>());
+            return (input >= 0) && isRepresentable<T>(static_cast<uint64_t>(input));
+        }
+        else
+        {
+            uint64_t input = 0;
+            PyArray_CastScalarToCtype(obj, &input, getNumpyTypeDescriptor<uint64_t>());
+            return isRepresentable<T>(input);
+        }
+        return false;
+    }
+}
+
+
+template<class T>
+bool parseNumpyScalar(PyObject* obj, T& value)
+{
+    if (PyArray_CheckScalar(obj))
+    {
+        // According to the numpy documentation:
+        // There are 21 statically-defined PyArray_Descr objects for the built-in data-types
+        PyArray_Descr* to = getNumpyTypeDescriptor<T>();
+        if (canBeSafelyCasted<T>(obj, to))
```

Comment:
> `canBeSafelyCasted<T>`

This looks over complicated. Code doesn't use all 21 types, because it is OpenCV code (I see only 3 specializations for `parseNumpyScalar` - `size_t`, `float`, `double`).

I suggest to remove "stdx" replacement and just write these specializations (or overloads).

Reply:
The underlying idea of this generalized approach is to reduce code repetition and have a set well tested functions as building blocks. Basically, it is possible to cover all primitive types by 3 general functions:

- For signed integral types;
- For unsigned integral types;
- For floating point types.

And 1 specialized function for `size_t`. 

After primitive all primitive types are covered by this set of functions, it is possible to cover other cases with their combination.
For instance,  `vector<type>` can be parsed as a sequence without fixed size, `cv::Size` can be parsed as a sequence with size 2. To parse a single element generalized function should be called.

If a new type (e.g. `int64_t` can not be parsed now) will be added, it automatically be covered by signed integral parser, so the contributor should provide only tests for it.

These functions are not introduced in the current PR, but will be.

If you are against this idea, I'm OK to rewrite this part with parsers specializations.

Reply:
1. Vector conversions is another story. I still didn't get problems with them.


2.
> If a new type

Set of types for bindings should be limited to cover other than PYthon languages too.

---

Lets keep patch as much simple as possible.

Reply:
@alalek 
Added some test to illustrate current problem with sequential types parsing (such as `vector` or `Size` (sequence of size 2)):
https://github.com/VadimLevin/opencv/commit/07bdc4ad6c6d86627d20cdb203c27d03e7ac8652

For `vector<int>` skipped test either fails with some inner error or perform narrowing conversion (`double` -> `int`) which we want to eliminate with strong type check.

For `cv::Size` it simply rejects everything except tuple.

So the main goal of the underlying idea, that I tried to describe in the previous comment, is elimination of confusing and non-intuitive conversion behavior.
Implementation with templates may be a bit complicated, but in my opinion has more advantages rather disadvantages.